src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -1,12 +1,80 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Concurrent;
 using System.Diagnostics.Contracts;
 using System;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Runtime
 {
+    /// <summary>
+    /// This class coalesces timeout tokens because cancelation tokens with timeouts are more expensive to expose.
+    /// Disposing too many such tokens will cause thread contentions in high throughput scenario.
+    ///
+    /// Tokens with target cancelation time 15ms apart would resolve to the same instance.
+    /// </summary>
+    static class TimeoutTokenSource
+    {
+        private const int COALESCING_SPAN_MS = 15;
+        private static readonly ConcurrentDictionary<long, Task<CancellationToken>> s_tokenCache =
+            new ConcurrentDictionary<long, Task<CancellationToken>>();
+
+        public static CancellationToken FromTimeout(int millisecondsTimeout)
+        {
+            return FromTimeoutAsync(millisecondsTimeout).Result;
+        }
+
+        public static Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout <= 0)
+            {
+                return Task.FromResult(new CancellationToken(true));
```

Comment:
Is it common for millisecondsTimeout to be 0?  If so, I'd suggest caching this task into a static, e.g.
```C#
private static readonly s_canceledTokenTask = Task.FromResult(new CancellationToken(true));
...
if (millisecondsTimeout <= 0)
{
    return s_canceledTokenTask;
}
```